File: microgpt.swift
/**
 * The most atomic way to train and inference a GPT in pure, dependency-free Swift.
 * This file is the complete algorithm.
 * Everything else is just efficiency.
 *
 * @karpathy (original Python), converted to Swift
 */

import Foundation

// --- RNG (xoshiro256** seeded with 42) ---
class Rng {
    var s: (UInt64, UInt64, UInt64, UInt64)

    init(seed: UInt64) {
        // SplitMix64 to seed xoshiro
        var sm = seed
        func next() -> UInt64 {
            sm = sm &+ 0x9e3779b97f4a7c15
            var z = sm
            z = (z ^ (z >> 30)) &* 0xbf58476d1ce4e5b9
            z = (z ^ (z >> 27)) &* 0x94d049bb133111eb
            return z ^ (z >> 31)
        }
        s = (next(), next(), next(), next())
    }

    func rotl(_ x: UInt64, _ k: Int) -> UInt64 {
        (x << k) | (x >> (64 - k))
    }

    func nextU64() -> UInt64 {
        let result = rotl(s.1 &* 5, 7) &* 9
        let t = s.1 << 17
        s.2 ^= s.0
        s.3 ^= s.1
        s.1 ^= s.2
        s.0 ^= s.3
        s.2 ^= t
        s.3 = rotl(s.3, 45)
        return result
    }

    func nextF64() -> Double {
        Double(nextU64() >> 11) / Double(1 << 53)
    }

    func gauss(_ mean: Double, _ std: Double) -> Double {
        let u1 = max(nextF64(), 1e-30)
        let u2 = nextF64()
        return mean + std * sqrt(-2.0 * log(u1)) * cos(2.0 * .pi * u2)
    }

    func shuffle<T>(_ array: inout [T]) {
        for i in stride(from: array.count - 1, through: 1, by: -1) {
            let j = Int(nextU64() % UInt64(i + 1))
            array.swapAt(i, j)
        }
    }

    func weightedChoice(_ weights: [Double]) -> Int {
        let total = weights.reduce(0, +)
        var r = nextF64() * total
        for (i, w) in weights.enumerated() {
            r -= w
            if r <= 0 { return i }
        }
        return weights.count - 1
    }
}

// --- Let there be Autograd ---
class Value {
    var data: Double                     // scalar value of this node calculated during forward pass
    var grad: Double = 0.0              // derivative of the loss w.r.t. this node
    var children: [Value]               // children of this node in the computation graph
    var localGrads: [Double]            // local derivative of this node w.r.t. its children

    init(_ data: Double, children: [Value] = [], localGrads: [Double] = []) {
        self.data = data
        self.children = children
        self.localGrads = localGrads
    }

    func vpow(_ exp: Double) -> Value {
        Value(pow(data, exp), children: [self], localGrads: [exp * pow(data, exp - 1)])
    }

    func vlog() -> Value {
        Value(log(data), children: [self], localGrads: [1.0 / data])
    }

    func vexp() -> Value {
        let e = exp(data)
        return Value(e, children: [self], localGrads: [e])
    }

    func relu() -> Value {
        Value(max(0, data), children: [self], localGrads: [data > 0 ? 1.0 : 0.0])
    }

    func backward() {
        var topo: [Value] = []
        var visited = Set<ObjectIdentifier>()

        func buildTopo(_ v: Value) {
            let id = ObjectIdentifier(v)
            if visited.contains(id) { return }
            visited.insert(id)
            for child in v.children {
                buildTopo(child)
            }
            topo.append(v)
        }

        buildTopo(self)
        self.grad = 1.0
        for v in topo.reversed() {
            for (child, lg) in zip(v.children, v.localGrads) {
                child.grad += lg * v.grad
            }
        }
    }
}

func + (lhs: Value, rhs: Value) -> Value {
    Value(lhs.data + rhs.data, children: [lhs, rhs], localGrads: [1.0, 1.0])
}
func + (lhs: Value, rhs: Double) -> Value { lhs + Value(rhs) }
func + (lhs: Double, rhs: Value) -> Value { Value(lhs) + rhs }

func * (lhs: Value, rhs: Value) -> Value {
    Value(lhs.data * rhs.data, children: [lhs, rhs], localGrads: [rhs.data, lhs.data])
}
func * (lhs: Value, rhs: Double) -> Value { lhs * Value(rhs) }
func * (lhs: Double, rhs: Value) -> Value { Value(lhs) * rhs }

prefix func - (v: Value) -> Value { v * (-1.0) }
func - (lhs: Value, rhs: Value) -> Value { lhs + (-rhs) }
func / (lhs: Value, rhs: Value) -> Value { lhs * rhs.vpow(-1.0) }
func / (lhs: Value, rhs: Double) -> Value { lhs * (1.0 / rhs) }

// --- Utility ---
typealias Vec1 = [Value]
typealias Mat = [Vec1]

func sumVals(_ v: [Value]) -> Value {
    var s = Value(0.0)
    for x in v { s = s + x }
    return s
}

// --- Model functions ---
func linear(_ x: Vec1, _ w: Mat) -> Vec1 {
    w.map { wo in
        var s = Value(0.0)
        for (wi, xi) in zip(wo, x) { s = s + wi * xi }
        return s
    }
}

func softmax(_ logits: Vec1) -> Vec1 {
    let maxVal = logits.map { $0.data }.max()!
    let exps = logits.map { ($0 - maxVal).vexp() }
    let total = sumVals(exps)
    return exps.map { $0 / total }
}

func rmsnorm(_ x: Vec1) -> Vec1 {
    var ms = Value(0.0)
    for xi in x { ms = ms + xi * xi }
    ms = ms / Double(x.count)
    let scale = (ms + 1e-5).vpow(-0.5)
    return x.map { $0 * scale }
}

// --- Main execution ---
let rng = Rng(seed: 42)

// Let there be an input dataset
guard let content = try? String(contentsOfFile: "input.txt", encoding: .utf8) else {
    print("Error: input.txt not found. Please download it first.")
    exit(1)
}
var docs = content.split(separator: "\n").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
rng.shuffle(&docs)
print("num docs: \(docs.count)")

// Let there be a Tokenizer
let uchars = Array(Set(docs.joined()).sorted())
let BOS = uchars.count
let vocabSize = uchars.count + 1
print("vocab size: \(vocabSize)")

func charIndex(_ c: Character) -> Int {
    uchars.firstIndex(of: c)!
}

// Initialize the parameters
let nEmbd = 16
let nHead = 4
let nLayer = 1
let blockSize = 16
let headDim = nEmbd / nHead

func makeMatrix(_ nout: Int, _ nin: Int) -> Mat {
    (0..<nout).map { _ in (0..<nin).map { _ in Value(rng.gauss(0, 0.08)) } }
}

let wte = makeMatrix(vocabSize, nEmbd)
let wpe = makeMatrix(blockSize, nEmbd)
let lmHead = makeMatrix(vocabSize, nEmbd)

struct LayerWeights {
    let attnWq: Mat
    let attnWk: Mat
    let attnWv: Mat
    let attnWo: Mat
    let mlpFc1: Mat
    let mlpFc2: Mat
}

var layers: [LayerWeights] = []
for _ in 0..<nLayer {
    layers.append(LayerWeights(
        attnWq: makeMatrix(nEmbd, nEmbd),
        attnWk: makeMatrix(nEmbd, nEmbd),
        attnWv: makeMatrix(nEmbd, nEmbd),
        attnWo: makeMatrix(nEmbd, nEmbd),
        mlpFc1: makeMatrix(4 * nEmbd, nEmbd),
        mlpFc2: makeMatrix(nEmbd, 4 * nEmbd)
    ))
}

// Flatten params
var params: [Value] = []
for row in wte { for p in row { params.append(p) } }
for row in wpe { for p in row { params.append(p) } }
for row in lmHead { for p in row { params.append(p) } }
for lw in layers {
    for mat in [lw.attnWq, lw.attnWk, lw.attnWv, lw.attnWo, lw.mlpFc1, lw.mlpFc2] {
        for row in mat { for p in row { params.append(p) } }
    }
}
print("num params: \(params.count)")

// GPT function
func gpt(_ tokenId: Int, _ posId: Int,
         _ keys: inout [[Vec1]], _ values: inout [[Vec1]]) -> Vec1 {
    let tokEmb = wte[tokenId]
    let posEmb = wpe[posId]
    var x: Vec1 = zip(tokEmb, posEmb).map { $0 + $1 }
    x = rmsnorm(x)

    for li in 0..<nLayer {
        // 1) Multi-head attention block
        let xResidual = x
        x = rmsnorm(x)
        let q = linear(x, layers[li].attnWq)
        let k = linear(x, layers[li].attnWk)
        let v = linear(x, layers[li].attnWv)
        keys[li].append(k)
        values[li].append(v)
        var xAttn: Vec1 = []
        for h in 0..<nHead {
            let hs = h * headDim
            let qH = Array(q[hs..<hs + headDim])
            let kH = keys[li].map { Array($0[hs..<hs + headDim]) }
            let vH = values[li].map { Array($0[hs..<hs + headDim]) }
            let scale = sqrt(Double(headDim))
            let attnLogits: Vec1 = kH.map { kt in
                var dot = Value(0.0)
                for j in 0..<headDim { dot = dot + qH[j] * kt[j] }
                return dot / scale
            }
            let attnWeights = softmax(attnLogits)
            for j in 0..<headDim {
                var s = Value(0.0)
                for t in 0..<vH.count { s = s + attnWeights[t] * vH[t][j] }
                xAttn.append(s)
            }
        }
        x = linear(xAttn, layers[li].attnWo)
        for i in 0..<nEmbd { x[i] = x[i] + xResidual[i] }
        // 2) MLP block
        let xResidual2 = x
        x = rmsnorm(x)
        x = linear(x, layers[li].mlpFc1)
        x = x.map { $0.relu() }
        x = linear(x, layers[li].mlpFc2)
        for i in 0..<nEmbd { x[i] = x[i] + xResidual2[i] }
    }

    return linear(x, lmHead)
}

// Let there be Adam
let learningRate = 0.01
let beta1 = 0.85
let beta2 = 0.99
let epsAdam = 1e-8
var mBuf = [Double](repeating: 0.0, count: params.count)
var vBuf = [Double](repeating: 0.0, count: params.count)

// Repeat in sequence
let numSteps = 1000
for step in 0..<numSteps {
    // Take single document, tokenize it
    let doc = docs[step % docs.count]
    var tokens = [BOS]
    for c in doc { tokens.append(charIndex(c)) }
    tokens.append(BOS)
    let n = min(blockSize, tokens.count - 1)

    // Forward
    var keys: [[Vec1]] = (0..<nLayer).map { _ in [] }
    var vals: [[Vec1]] = (0..<nLayer).map { _ in [] }
    var losses: Vec1 = []
    for posId in 0..<n {
        let tokenId = tokens[posId]
        let targetId = tokens[posId + 1]
        let logits = gpt(tokenId, posId, &keys, &vals)
        let probs = softmax(logits)
        let lossT = -(probs[targetId].vlog())
        losses.append(lossT)
    }
    let loss = sumVals(losses) / Double(n)

    // Backward
    loss.backward()

    // Adam optimizer update
    let lrT = learningRate * (1.0 - Double(step) / Double(numSteps))
    for (i, p) in params.enumerated() {
        mBuf[i] = beta1 * mBuf[i] + (1 - beta1) * p.grad
        vBuf[i] = beta2 * vBuf[i] + (1 - beta2) * p.grad * p.grad
        let mHat = mBuf[i] / (1.0 - pow(beta1, Double(step + 1)))
        let vHat = vBuf[i] / (1.0 - pow(beta2, Double(step + 1)))
        p.data -= lrT * mHat / (sqrt(vHat) + epsAdam)
        p.grad = 0
    }

    print(String(format: "step %4d / %4d | loss %.4f", step + 1, numSteps, loss.data))
}

// Inference
let temperature = 0.5
print("\n--- inference (new, hallucinated names) ---")
for sampleIdx in 0..<20 {
    var keys: [[Vec1]] = (0..<nLayer).map { _ in [] }
    var vals: [[Vec1]] = (0..<nLayer).map { _ in [] }
    var tokenId = BOS
    var sample: [Character] = []
    for posId in 0..<blockSize {
        let logits = gpt(tokenId, posId, &keys, &vals)
        let tempLogits = logits.map { $0 / temperature }
        let probs = softmax(tempLogits)
        let weights = probs.map { $0.data }
        tokenId = rng.weightedChoice(weights)
        if tokenId == BOS { break }
        sample.append(uchars[tokenId])
    }
    print(String(format: "sample %2d: %@", sampleIdx + 1, String(sample)))
}

File: microgpt.go
// The most atomic way to train and inference a GPT in pure, dependency-free Go.
// This file is the complete algorithm.
// Everything else is just efficiency.
//
// @karpathy (original Python), converted to Go

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
)

// --- RNG (xoshiro256** seeded with 42) ---
type Rng struct {
	s [4]uint64
}

func NewRng(seed uint64) *Rng {
	r := &Rng{}
	sm := seed
	smNext := func() uint64 {
		sm += 0x9e3779b97f4a7c15
		z := sm
		z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9
		z = (z ^ (z >> 27)) * 0x94d049bb133111eb
		return z ^ (z >> 31)
	}
	r.s[0] = smNext()
	r.s[1] = smNext()
	r.s[2] = smNext()
	r.s[3] = smNext()
	return r
}

func rotl(x uint64, k uint) uint64 {
	return (x << k) | (x >> (64 - k))
}

func (r *Rng) NextU64() uint64 {
	result := rotl(r.s[1]*5, 7) * 9
	t := r.s[1] << 17
	r.s[2] ^= r.s[0]
	r.s[3] ^= r.s[1]
	r.s[1] ^= r.s[2]
	r.s[0] ^= r.s[3]
	r.s[2] ^= t
	r.s[3] = rotl(r.s[3], 45)
	return result
}

func (r *Rng) NextF64() float64 {
	return float64(r.NextU64()>>11) / float64(uint64(1)<<53)
}

func (r *Rng) Gauss(mean, std float64) float64 {
	u1 := math.Max(r.NextF64(), 1e-30)
	u2 := r.NextF64()
	return mean + std*math.Sqrt(-2.0*math.Log(u1))*math.Cos(2.0*math.Pi*u2)
}

func (r *Rng) Shuffle(n int, swap func(i, j int)) {
	for i := n - 1; i > 0; i-- {
		j := int(r.NextU64() % uint64(i+1))
		swap(i, j)
	}
}

func (r *Rng) WeightedChoice(weights []float64) int {
	total := 0.0
	for _, w := range weights {
		total += w
	}
	rv := r.NextF64() * total
	for i, w := range weights {
		rv -= w
		if rv <= 0 {
			return i
		}
	}
	return len(weights) - 1
}

// --- Let there be Autograd ---
type Value struct {
	Data       float64
	Grad       float64
	children   []*Value
	localGrads []float64
}

func NewValue(data float64, children []*Value, localGrads []float64) *Value {
	return &Value{Data: data, Grad: 0, children: children, localGrads: localGrads}
}

func Val(d float64) *Value {
	return NewValue(d, nil, nil)
}

func Add(a, b *Value) *Value {
	return NewValue(a.Data+b.Data, []*Value{a, b}, []float64{1, 1})
}

func AddScalar(a *Value, s float64) *Value {
	return Add(a, Val(s))
}

func Mul(a, b *Value) *Value {
	return NewValue(a.Data*b.Data, []*Value{a, b}, []float64{b.Data, a.Data})
}

func MulScalar(a *Value, s float64) *Value {
	return Mul(a, Val(s))
}

func Pow(a *Value, exp float64) *Value {
	return NewValue(math.Pow(a.Data, exp), []*Value{a}, []float64{exp * math.Pow(a.Data, exp-1)})
}

func Log(a *Value) *Value {
	return NewValue(math.Log(a.Data), []*Value{a}, []float64{1.0 / a.Data})
}

func Exp(a *Value) *Value {
	e := math.Exp(a.Data)
	return NewValue(e, []*Value{a}, []float64{e})
}

func Relu(a *Value) *Value {
	d := 0.0
	g := 0.0
	if a.Data > 0 {
		d = a.Data
		g = 1.0
	}
	return NewValue(d, []*Value{a}, []float64{g})
}

func Neg(a *Value) *Value    { return MulScalar(a, -1) }
func Sub(a, b *Value) *Value { return Add(a, Neg(b)) }
func Div(a, b *Value) *Value { return Mul(a, Pow(b, -1)) }
func DivScalar(a *Value, s float64) *Value {
	return MulScalar(a, 1.0/s)
}

func Backward(root *Value) {
	topo := make([]*Value, 0)
	visited := make(map[*Value]bool)
	var buildTopo func(v *Value)
	buildTopo = func(v *Value) {
		if visited[v] {
			return
		}
		visited[v] = true
		for _, child := range v.children {
			buildTopo(child)
		}
		topo = append(topo, v)
	}
	buildTopo(root)
	root.Grad = 1.0
	for i := len(topo) - 1; i >= 0; i-- {
		v := topo[i]
		for j, child := range v.children {
			child.Grad += v.localGrads[j] * v.Grad
		}
	}
}

// --- Utility ---
func SumVals(v []*Value) *Value {
	s := Val(0)
	for _, x := range v {
		s = Add(s, x)
	}
	return s
}

// --- Model functions ---
func Linear(x []*Value, w [][]*Value) []*Value {
	out := make([]*Value, len(w))
	for i, wo := range w {
		s := Val(0)
		for j := range wo {
			s = Add(s, Mul(wo[j], x[j]))
		}
		out[i] = s
	}
	return out
}

func Softmax(logits []*Value) []*Value {
	maxVal := logits[0].Data
	for _, v := range logits[1:] {
		if v.Data > maxVal {
			maxVal = v.Data
		}
	}
	exps := make([]*Value, len(logits))
	for i, v := range logits {
		exps[i] = Exp(AddScalar(v, -maxVal))
	}
	total := SumVals(exps)
	out := make([]*Value, len(exps))
	for i, e := range exps {
		out[i] = Div(e, total)
	}
	return out
}

func Rmsnorm(x []*Value) []*Value {
	ms := Val(0)
	for _, xi := range x {
		ms = Add(ms, Mul(xi, xi))
	}
	ms = DivScalar(ms, float64(len(x)))
	scale := Pow(AddScalar(ms, 1e-5), -0.5)
	out := make([]*Value, len(x))
	for i, xi := range x {
		out[i] = Mul(xi, scale)
	}
	return out
}

// --- Layer weights ---
type LayerWeights struct {
	attnWq, attnWk, attnWv, attnWo [][]*Value
	mlpFc1, mlpFc2                  [][]*Value
}

func main() {
	rng := NewRng(42)

	// Let there be an input dataset
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error: input.txt not found. Please download it first.")
		return
	}
	defer file.Close()

	var docs []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			docs = append(docs, line)
		}
	}
	rng.Shuffle(len(docs), func(i, j int) { docs[i], docs[j] = docs[j], docs[i] })
	fmt.Printf("num docs: %d\n", len(docs))

	// Let there be a Tokenizer
	charSet := make(map[rune]bool)
	for _, d := range docs {
		for _, c := range d {
			charSet[c] = true
		}
	}
	uchars := make([]rune, 0, len(charSet))
	for c := range charSet {
		uchars = append(uchars, c)
	}
	sort.Slice(uchars, func(i, j int) bool { return uchars[i] < uchars[j] })
	bos := len(uchars)
	vocabSize := len(uchars) + 1
	fmt.Printf("vocab size: %d\n", vocabSize)

	charIndex := func(c rune) int {
		for i, u := range uchars {
			if u == c {
				return i
			}
		}
		return -1
	}

	// Initialize the parameters
	nEmbd := 16
	nHead := 4
	nLayer := 1
	blockSize := 16
	headDim := nEmbd / nHead

	makeMatrix := func(nout, nin int) [][]*Value {
		m := make([][]*Value, nout)
		for i := range m {
			m[i] = make([]*Value, nin)
			for j := range m[i] {
				m[i][j] = Val(rng.Gauss(0, 0.08))
			}
		}
		return m
	}

	wte := makeMatrix(vocabSize, nEmbd)
	wpe := makeMatrix(blockSize, nEmbd)
	lmHead := makeMatrix(vocabSize, nEmbd)

	layers := make([]LayerWeights, nLayer)
	for i := range layers {
		layers[i] = LayerWeights{
			attnWq: makeMatrix(nEmbd, nEmbd),
			attnWk: makeMatrix(nEmbd, nEmbd),
			attnWv: makeMatrix(nEmbd, nEmbd),
			attnWo: makeMatrix(nEmbd, nEmbd),
			mlpFc1: makeMatrix(4*nEmbd, nEmbd),
			mlpFc2: makeMatrix(nEmbd, 4*nEmbd),
		}
	}

	// Flatten params
	var params []*Value
	for _, row := range wte {
		params = append(params, row...)
	}
	for _, row := range wpe {
		params = append(params, row...)
	}
	for _, row := range lmHead {
		params = append(params, row...)
	}
	for _, lw := range layers {
		for _, mat := range [][][]*Value{lw.attnWq, lw.attnWk, lw.attnWv, lw.attnWo, lw.mlpFc1, lw.mlpFc2} {
			for _, row := range mat {
				params = append(params, row...)
			}
		}
	}
	fmt.Printf("num params: %d\n", len(params))

	// GPT function
	gpt := func(tokenID, posID int, keys, values [][][]*Value) []*Value {
		tokEmb := wte[tokenID]
		posEmb := wpe[posID]
		x := make([]*Value, nEmbd)
		for i := range x {
			x[i] = Add(tokEmb[i], posEmb[i])
		}
		x = Rmsnorm(x)

		for li := 0; li < nLayer; li++ {
			// 1) Multi-head attention block
			xResidual := make([]*Value, nEmbd)
			copy(xResidual, x)
			x = Rmsnorm(x)
			q := Linear(x, layers[li].attnWq)
			k := Linear(x, layers[li].attnWk)
			v := Linear(x, layers[li].attnWv)
			keys[li] = append(keys[li], k)
			values[li] = append(values[li], v)
			xAttn := make([]*Value, 0, nEmbd)
			for h := 0; h < nHead; h++ {
				hs := h * headDim
				qH := q[hs : hs+headDim]
				kH := make([][]*Value, len(keys[li]))
				vH := make([][]*Value, len(values[li]))
				for t := range keys[li] {
					kH[t] = keys[li][t][hs : hs+headDim]
					vH[t] = values[li][t][hs : hs+headDim]
				}
				scale := math.Sqrt(float64(headDim))
				attnLogits := make([]*Value, len(kH))
				for t := range kH {
					dot := Val(0)
					for j := 0; j < headDim; j++ {
						dot = Add(dot, Mul(qH[j], kH[t][j]))
					}
					attnLogits[t] = DivScalar(dot, scale)
				}
				attnWeights := Softmax(attnLogits)
				for j := 0; j < headDim; j++ {
					s := Val(0)
					for t := range vH {
						s = Add(s, Mul(attnWeights[t], vH[t][j]))
					}
					xAttn = append(xAttn, s)
				}
			}
			x = Linear(xAttn, layers[li].attnWo)
			for i := 0; i < nEmbd; i++ {
				x[i] = Add(x[i], xResidual[i])
			}
			// 2) MLP block
			xResidual2 := make([]*Value, nEmbd)
			copy(xResidual2, x)
			x = Rmsnorm(x)
			x = Linear(x, layers[li].mlpFc1)
			for i := range x {
				x[i] = Relu(x[i])
			}
			x = Linear(x, layers[li].mlpFc2)
			for i := 0; i < nEmbd; i++ {
				x[i] = Add(x[i], xResidual2[i])
			}
		}

		return Linear(x, lmHead)
	}

	// Let there be Adam
	learningRate := 0.01
	beta1 := 0.85
	beta2 := 0.99
	epsAdam := 1e-8
	mBuf := make([]float64, len(params))
	vBuf := make([]float64, len(params))

	// Repeat in sequence
	numSteps := 1000
	for step := 0; step < numSteps; step++ {
		// Take single document, tokenize it
		doc := docs[step%len(docs)]
		tokens := []int{bos}
		for _, c := range doc {
			tokens = append(tokens, charIndex(c))
		}
		tokens = append(tokens, bos)
		n := blockSize
		if len(tokens)-1 < n {
			n = len(tokens) - 1
		}

		// Forward
		keys := make([][][]*Value, nLayer)
		vals := make([][][]*Value, nLayer)
		for i := range keys {
			keys[i] = make([][]*Value, 0)
			vals[i] = make([][]*Value, 0)
		}
		var losses []*Value
		for posID := 0; posID < n; posID++ {
			tokenID := tokens[posID]
			targetID := tokens[posID+1]
			logits := gpt(tokenID, posID, keys, vals)
			probs := Softmax(logits)
			lossT := Neg(Log(probs[targetID]))
			losses = append(losses, lossT)
		}
		loss := DivScalar(SumVals(losses), float64(n))

		// Backward
		Backward(loss)

		// Adam optimizer update
		lrT := learningRate * (1.0 - float64(step)/float64(numSteps))
		for i, p := range params {
			mBuf[i] = beta1*mBuf[i] + (1-beta1)*p.Grad
			vBuf[i] = beta2*vBuf[i] + (1-beta2)*p.Grad*p.Grad
			mHat := mBuf[i] / (1.0 - math.Pow(beta1, float64(step+1)))
			vHat := vBuf[i] / (1.0 - math.Pow(beta2, float64(step+1)))
			p.Data -= lrT * mHat / (math.Sqrt(vHat) + epsAdam)
			p.Grad = 0
		}

		fmt.Printf("step %4d / %4d | loss %.4f\n", step+1, numSteps, loss.Data)
	}

	// Inference
	temperature := 0.5
	fmt.Println("\n--- inference (new, hallucinated names) ---")
	for sampleIdx := 0; sampleIdx < 20; sampleIdx++ {
		keys := make([][][]*Value, nLayer)
		vals := make([][][]*Value, nLayer)
		for i := range keys {
			keys[i] = make([][]*Value, 0)
			vals[i] = make([][]*Value, 0)
		}
		tokenID := bos
		var sample []rune
		for posID := 0; posID < blockSize; posID++ {
			logits := gpt(tokenID, posID, keys, vals)
			tempLogits := make([]*Value, len(logits))
			for i, l := range logits {
				tempLogits[i] = DivScalar(l, temperature)
			}
			probs := Softmax(tempLogits)
			weights := make([]float64, len(probs))
			for i, p := range probs {
				weights[i] = p.Data
			}
			tokenID = rng.WeightedChoice(weights)
			if tokenID == bos {
				break
			}
			sample = append(sample, uchars[tokenID])
		}
		fmt.Printf("sample %2d: %s\n", sampleIdx+1, string(sample))
	}
}
